e2store/src/era2.rs

```diff
@@ -0,0 +1,313 @@
+        })
+    }
+
+    pub fn create(era2_path: PathBuf, header: Header) -> anyhow::Result<Self> {
+        if era2_path.is_file() {
+            panic!(
+                "era2_path is not a directory, it is a file: {:?}",
+                era2_path
+            );
+        }
+        let era2_path = era2_path.join(format!(
+            "mainnet-{:010}-{}.era2",
+            header.number,
+            hex::encode(&header.state_root.as_slice()[..4])
+        ));
+        if era2_path.exists() {
+            panic!("era2 file already exists: {:?}", era2_path);
+        }
+        let mut e2store_stream = E2StoreStream::create(&era2_path)?;
+
+        let version = VersionEntry::default();
+        e2store_stream.append_entry(&version.clone().into())?;
+
+        let header = HeaderEntry { header };
+        e2store_stream.append_entry(&header.clone().try_into()?)?;
+
+        Ok(Self {
+            version,
+            header,
+            e2store_stream,
+            storage_entries_left_to_decode: 0,
+        })
+    }
+
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> anyhow::Result<Era2StreamEntry> {
+        if self.storage_entries_left_to_decode > 0 {
+            self.storage_entries_left_to_decode -= 1;
+            return Ok(Era2StreamEntry::Storage(StorageEntry::try_from(
+                &self.e2store_stream.next_entry()?,
+            )?));
+        }
+        let account = AccountEntry::try_from(&self.e2store_stream.next_entry()?)?;
+        self.storage_entries_left_to_decode = account.storage_count;
+        Ok(Era2StreamEntry::Account(account))
+    }
+
+    pub fn append_entry(&mut self, entry: &Era2StreamEntry) -> anyhow::Result<usize> {
+        let size = match entry {
+            Era2StreamEntry::Account(account) => {
+                if self.storage_entries_left_to_decode != 0 {
+                    panic!("Invalid append entry state: expected a storage entry, got an account entry. Still have {} storage entries left to append", self.storage_entries_left_to_decode);
+                }
+
+                self.storage_entries_left_to_decode = account.storage_count;
+                let entry: Entry = account.clone().try_into()?;
+                self.e2store_stream.append_entry(&entry)?;
+                entry.value.len()
+            }
+            Era2StreamEntry::Storage(storage) => {
+                if self.storage_entries_left_to_decode == 0 {
+                    panic!("Invalid append entry state: expected an account entry, got a storage entry. No storage entries left to append for the account");
+                }
+
+                self.storage_entries_left_to_decode -= 1;
+                let entry: Entry = storage.clone().try_into()?;
+                self.e2store_stream.append_entry(&entry)?;
+                entry.value.len()
+            }
+        };
+        Ok(size)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Era2StreamEntry {
```

Comment:
> I'm also fine with leaving it as is, if you have strong preferences for this one.

I wouldn't say I have strong preferences on this. I also don't like the current name.

> How about we rename this to AccountEntry and rename current AccountEntry to AccountHeaderEntry

I don't like this because the current `AccountEntry` isn't really a header so it feels confusing

I will change `Era2StreamEntry` to `StreamEntry` for now. I don't really like `StreamEntry` too, if a good name is found I would be happy to rename it.

Reply:
Since neither of us likes `StreamEntry`, I will give few more suggestions but feel free to leave it as is if you don't like them:

1. rename `StreamEntry` to `AccountOrStorageEntry` (probably my favorite)
2. rename `StreamEntry` to `AccountEntry` and `AccountEntry` to `AccountInfoEntry` (since you don't like header)

Reply:
I think suggestion 1 is good so I will update it to that